src/buildstream/_cas/cascache.py

```diff
@@ -443,6 +443,48 @@ def import_directory(self, path: str, properties: Optional[List[str]] = None) ->
 
         return utils._message_digest(root_directory)
 
+    # stage_directory():
+    #
+    # A contextmanager to stage a CAS directory tree in the local filesystem.
+    #
+    # This makes the specified directory tree temporarily available for local
+    # filesystem access. This may use FUSE or hardlinking.
```

Comment:
So if I understand correctly, this patch *may* use `buildbox-fuse` if available to create the local filesystem access required for a plugin to go and do it's work with previous sources staged.

What worries me here is *"This may use FUSE or hardlinking."*, is it possible that this uses hardlinking *without* FUSE ?

We should avoid a scenario where we expose writable hardlinks into the local CAS, to a source plugin which *may* in some unlikely even modify a staged file (instead of replacing/re-installing a new one).

Reply:
`StageTree` is the same method that is also used by buildbox-run-bubblewrap. I.e. this shouldn't add any risk. For safe hardlinking, buildbox-casd has to run as a different user, in which case this should be safe for both build sandboxes and sources.

That said, I don't remember whether we already verify that the setup is correct.

Reply:
One thing that is not obvious, is that this `StageTree` being safe for running commands inside a sandbox using buildbox-run is equally safe for us to go right ahead and run buildstream subprocesses. I would not have made that assumption.

I of course believe you that it is safe, just saying that it doesn't seem obvious that it *would* be safe :)